rfcs/2022-07-28-13691-registered-internal-events.md

```diff
@@ -0,0 +1,231 @@
+to do a little magic with event names in tests).
+
+```rust
+emit!(BytesSent {
+    byte_size: 12345,
+    protocol: "https",
+});
+```
+
+To continue the same pattern, we will set up a new trait to register an event. As above, it will
+consume the event data (and so take ownership of all the fields) and return the handle. This new
+return type for the event handle will necessarily be named by the trait.
+
+```rust
+trait RegisterInternalEvent: Sized {
+    type Handle: InternalEventHandle;
+    fn register(self) -> Self::Handle;
+}
+```
+
+Since we want to continue the functional conveniences of `emit`, we will also set up a function and
+macro wrapper to provide the convenience:
+
+```rust
+fn register<T: RegisterInternalEvent>(event: E) -> E::Handle {
+    event.register()
+}
+
+macro_rules! register {
+    ($event:expr) => { vector_core::internal_event::register($event) };
+}
+```
+
+We want the same test handling magic, so that will come included with its own wrapper:
+
+```rust
+struct DefaultHandleName<E> {
+    pub name: &'static str,
+    pub event: E,
+}
+
+#[cfg(test)]
+macro_rules! register {
+    ($event:expr) => {
+        vector_core::internal_event::register(
+            vector_core::internal_event::DefaultHandleName {
+                event: $event,
+                name: stringify!($event),
+            }
+        )
+    };
+}
+```
+
+The registered event handle will have its own trait for emitting the event, which enforces all such
+events follow exactly the same pattern. The input data is consumed, just as it is for the existing
+`emit` function.
+
+```rust
+trait InternalEventHandle {
+    type Data;
+    fn emit(&self, data: Data);
+}
+```
+
+To assist with providing the necessary types for emitting the registered events, a set of common
+unit structures will be defined for the required data:
+
+```rust
+struct ByteSize(usize);
+
+struct ByteSizeCount(usize, usize);
+```
+
+Finally, here is a sample implementation of the above traits, taken from the existing
+`EndpointBytesReceived` internal event:
+
+```rust
+use metrics::Counter;
+
+struct RegisterEndpointBytesReceived {
+    bytes_total: Counter,
+    protocol: &'static str,
+    endpoint: String,
+}
+
+struct EndpointBytesReceivedHandle {
+    bytes_total: Counter,
+    protocol: &'static str,
+    endpoint: String,
+}
+
+impl RegisterInternalEvent for RegisterEndpointBytesReceived {
+    type Handle = EndpointBytesReceivedHandle;
+
+    fn register(self) -> Self::Handle {
+        let bytes_total = register_counter!(
+            "component_received_bytes_total",
+            "protocol" => self.protocol,
+            "endpoint" => self.endpoint.clone(),
+        );
+        Self {
+            bytes_total,
+            protocol: self.protocol,
+            endpoint: self.endpoint,
+        }
+    }
+}
+
+impl InternalEventHandle for EndpointBytesReceivedHandle {
+    type Data = ByteSize;
+    fn emit(&self, data: ByteSize) {
```

Comment:
Or, more simply, `D: Into<Self::Data>`.

Yes, this might simplify some use cases, but we'll have to see if that becomes an issue. We don't want to have a situation where we emit multiple events which each call, say, `ByteSizeOf` which ends up recursing, etc.

Reply:
> Or, more simply, `D: Into<Self::Data>`.

There was something I was doing recently in the config schema work where using `Into` in that exact way lead to an "ambiguous type/conflicting implementations" sort of error message from the compiler.. but using `From` instead made it go away... I think related to reflexivity when `D` actually _was_ `Self::Data`, but anyways, you're not wrong and my problem might not actually apply here. :)